Add support for 512 character fonts on MAX7456, add boot logo

fiam opened the pull request:
- Add support for read/write MAX7456 NVM for chars > 256
- Refactor and cleanup the MAX7456 driver
- Add support for characters > 255 in the `display` driver layer
- Fix an small issue with the initialisation order of the display/max7456 drivers

Configurator PR https://github.com/iNavFlight/inav-configurator/pull/625

fiam committed:
Refactor code for locking access to MAX7456 to functions

Rename max7456Lock var to max7455Mutex. Add max7456Lock(),
max7456Unlock() and max7456TryLock().

fiam committed:
Add support for read/write MAX7456 NVM for chars > 256

Add max7456Character_t, which represents the data for character a
single MAX7456 character.
Use max7456Character_t to read and pass character data around.
Change max7456WriteNvm() to accept a 2-byte character address.
Change MSP handler for MSP_OSD_CHAR_WRITE to accept 2 byte character
addresses. To keep backwards compatibility, check wether the caller
sent 55 or 56 bytes.
Add max7456ReadNvm() to read a character from the MAX7456 NVM, which
will eventually be used for font metadata.

fiam committed:
Add workaround for corrupted characters when uploading MAX7456 fonts

Restore the previous behavior of rebooting after uploading characters,
but add a comment explaining why it needs to be done.

fiam committed:
Add support for OSD characters > 255, refactor MAX7456

Change display(Read|Write)Char* functions to use an uint16_t
character. This lets callers which need characters > 255 use these
functions, while keeping most strings in ASCII to save memory (
the string based functions still take a cosnt char *).

Change display drivers to support 16 bit characters when writing
a single character. Note that some drivers might silently truncate
the character, it's the responsability of the caller to check
wether the display supports characters > 255 (API will follow in
the next commit).

Refactor and cleanup the MAX7456 driver:

- Move global state to a struct, for better encapsulation
- Make max7456DrawScreenPartial() static, expose instead the new
function max7456Update(), which calls max7456StallCheck() and
max7456DrawScreenPartial() instead of doing all the work on
a single function.
- Rewrite max7456RefreshAll() to use HW based screen clearing, then
leverage max7456DrawScreenPartial() to redraw only non-blank
characters. This should make full screen redraws faster while
reducing complexity and flash usage (we no longer have 2 paths
for drawing).
- Support writing characters > 255, switching to 8bit mode on
demand.

fiam committed:
Add support for font metadata, use it for detecting extended fonts

Also, add boot logo on OSD when the font supports it. If the font
is invalid or out of date, show an "INVALID FONT" message.

fiam committed:
Call max7456Init() before displayInit()

Otherwise the display initialization might call methods on the
max7456 driver on an uninitialized device

fiam committed:
Ensure the MAX7456 detects a valid signal before accepting it

Code was not checking for the Loss-of-Sync bit, causing it to
detect PAL too early in some cases.

digitalentity commented on src/main/drivers/display.c:
```diff
@@ -214,6 +221,14 @@ uint16_t displayTxBytesFree(const displayPort_t *instance)
     return instance->vTable->txBytesFree(instance);
 }
 
+bool displayFontMetadata(displayFontMetadata_t *metadata, const displayPort_t *instance)
```
Maybe `displayGetFontMetadata`?

fiam replied:
Just this function or also the field in the vtable?

digitalentity replied:
The field as well - let's make names self-explanatory.

digitalentity commented on src/main/drivers/display.c:
```diff
@@ -157,26 +155,35 @@ int displayWriteWithAttr(displayPort_t *instance, uint8_t x, uint8_t y, const ch
 }
 
-int displayWriteChar(displayPort_t *instance, uint8_t x, uint8_t y, uint8_t c)
+int displayWriteChar(displayPort_t *instance, uint8_t x, uint8_t y, uint16_t c)
```
Maybe we should define `displaySymbol_t` type?

fiam replied:
I think it's not worth the complexity of having to look up what `displaySymbol_t` is. What concerns me about `uint16_t` is that it might make future developers think that any value up to `(2^16) - 1` is accepted. Is there any argument attribute that we can apply to tell the compiler the maximum size (so it will produce a warning when it gets implicitly truncated)?

digitalentity commented on src/main/drivers/max7456.c:
```diff
@@ -327,110 +428,135 @@ void max7456Write(uint8_t x, uint8_t y, const char *buff, uint8_t mode)
-            }
+            bufPtr = max7456PrepareBuffer(spiBuff, bufPtr, MAX7456ADD_DMAH, ph | DMAH_8_BIT_DMDI_IS_CHAR_ATTR);
+            bufPtr = max7456PrepareBuffer(spiBuff, bufPtr, MAX7456ADD_DMAL, pl);
+            bufPtr = max7456PrepareBuffer(spiBuff, bufPtr, MAX7456ADD_DMDI, charMode << 2);
```
It's not obvious that `charMode << 2` is actually referring to extra bit of 16-bit character. We should probably need to comment on this.

fiam replied:
It definitely needs a comment, but the shift not because of the extra bit for 9 bit chars. The reason is that we're storing that byte in `DMM` format, but when writing attributes via `DMDI` the bits are 2 positions up.

digitalentity replied:
So you see - I missed that point 😄

fiam committed:
Refactor the rest of the MAX7456 driver

- Cleanup init sequence to perform as much initialization as
possible in max7456Init().
- Initialize the OSD as soon as video sync is detected. If no
video sync is detected at 1.5s, start as PAL but then continue
probing until we find a valid video signal.
- Wait for the software reset to finish during initialization.
- Remove extern variable for maxScreenSize, replace it with
max7456GetScreenSize().
- Make sure that when disabling the OSD for reading/writing to
NVM we only enable it again if it was enabled before.
- Rewrite prototypes in the header file to use the same style and
formatting as the rest of the code.
- Delay call to millis() in max7456StallCheck() to avoid calling
it when it's not strictly needed.

fiam committed:
Rename displayFontMetadata() to displayGetFontMetadata()

fiam committed:
Add comment explaining the shift by 2 bits when writing to DMDI

fiam committed:
Rename fontMetadata in display driver's vTable to getFontMetadata
